Docs: CaaSP cluster management (#307)
uyuni-project/uyuni-docs: modules/client-configuration/pages/vhm-caasp.adoc

The management node cannot be part of the cluster, and it must have the {caasp} channels associated with it before you begin.
You can use a single management node for multiple clusters, as long as the clusters are all of the same kind.

.Procedure: Electing a Management Node
. In the {productname} {webui}, navigate to menu:Systems[System List] and click the name of the client to elect as the management node.
. Navigate to the menu:Formulas[Configuration] tab, and check the ``CaaSP Management Node`` formula.
. Click btn:[Save] and apply the highstate.

[NOTE]
====
You will not be able to use the management node until the highstate has been completed.
====

List all known clusters by navigating to menu:Clusters[Overview].
This list displays all existing clusters, along with the cluster type, and which management node they are associated with.
It also shows the nodes within the cluster, if the nodes are registered to {productname}.
For the nodes within a cluster, additional information from ``skuba`` and the {k8s} API are shown, including the role, status, and whether any updates are available.

> For more information about the data available for nodes, see https://documentation.suse.com/suse-caasp/4/html/caasp-admin/\_cluster_updates.html.

Reviewer: ```suggestion
For more information about the data available for nodes, see https://documentation.suse.com/suse-caasp/4/html/caasp-admin/_cluster_updates.html.
```

=== Manage Clusters

To manage a cluster in {productname}, add the cluster in the {webui}.

.Procedure: Adding an Existing Cluster
. In the {productname} {webui}, navigate to menu:Clusters[Overview] and click btn:[FIXME].
. Follow the prompts to provide information about your cluster, including the cluster type, and select the management node to associate.
. Type the path to the ``skuba`` configuration file for the cluster.
// For example?

> . Type the path passwordless SSH key you want to use, or to the ``ssh-agent`` socket.

Reviewer: ```suggestion
. Type the passwordless SSH key you want to use, or to the ``ssh-agent`` socket.
```

By default, the system group is called ``Cluster <cluster_name>``.
Refresh the system group to update the list of nodes.
Only nodes known to {productname} are shown.

You can remove clusters from {productname} by navigating to menu:Clusters[Overview], unchecking the cluster to be deleted, and clicking btn:[Delete Cluster].

[IMPORTANT]
====
Deleting a cluster removes the cluster from {productname}, it does not delete the cluster nodes.
Workloads running on the cluster will continue uninterrupted.
====

=== Manage Nodes

When you have the cluster created in {productname}, you can manage nodes within the cluster.

Before you add a new node to the cluster, check the management node can access the node you want to add using passwordless SSH, or the ``ssh-agent`` socket you are forwarding.

You also need to ensure that the node you want to add is registered to {productname}, and has a {caasp} channel assigned.

.Procedure: Adding a Node to a Cluster
. In the {productname} {webui}, navigate to menu:Clusters[Overview] and click btn:[Join Node].
. Select the node to add from a list of available nodes.

> The available nodes includes only nodes that are registered to {productname}, are not management nodes, and are not currently part of any cluster.

Reviewer: ```suggestion
The list of available nodes includes only nodes that are registered to {productname}, are not management nodes, and are not currently part of any cluster.
```

. Follow the prompts to enter the {caasp} parameters for the node to be added.
. OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the nodes that are being added.
. Click btn:[Save] to schedule an action to add the node.
During this action, {productname} prepares the node for joining by disabling swap, then joins the node to the cluster.

.Procedure: Removing a Node from a Cluster

> . In the {productname} {webui}, navigate to menu:Clusters[Overview], uncheck the nodes to remove, and click btn:[Remove Node].

Reviewer: ```suggestion
. In the {productname} {webui}, navigate to menu:Clusters[Overview], uncheck the node to remove, and click btn:[Remove Node].
```

. Follow the prompts to define the parameters for the node to be removed.

> . OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the nodes that are being removed.

Reviewer: ```suggestion
. OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the node that is being removed.
```

. Click btn:[Save] to schedule an action to remove the node.

> For more information about node removal, see https://documentation.suse.com/suse-caasp/4/single-html/caasp-admin/\#\_permanent_removal

Reviewer: ```suggestion
For more information about node removal, see https://documentation.suse.com/suse-caasp/4/single-html/caasp-admin/#_permanent_removal.
```
It would be weird if you actually must escape these characters

Reviewer 2: Oh, I escaped them because it was breaking my highlighting in my editor. Forgot to remove them, clearly!

==== Upgrade the Cluster

If the cluster has available updates, you can use {productname} to schedule and manage the upgrade.

{productname} upgrades all control planes first, and then upgrades the workers.

> For more information, see https://documentation.suse.com/suse-caasp/4.2/single-html/caasp-admin/\#\_cluster_updates

Reviewer: ```suggestion
For more information, see https://documentation.suse.com/suse-caasp/4.2/single-html/caasp-admin/#_cluster_updates.
```

    <addons config:type="list">
      <addon>
        <name>sle-module-containers</name>
        <version>15.1</version>
        <arch>x86_64</arch>
      </addon>
      <addon>
        <name>caasp</name>
        <version>4.0</version>
        <arch>x86_64</arch>
        <reg_code><!-- replace this comment with a CaaSP registration code --></reg_code>
      </addon>
    </addons>
  </suse_register>

  <scripts>
  <chroot-scripts config:type="list">
      <script>
        <chrooted config:type="boolean">true</chrooted>
        <filename>add_sles_sudo_rule.sh</filename>
        <interpreter>shell</interpreter>
        <source>
<![CDATA[
#!/bin/sh
echo "Defaults:sles !targetpw
sles ALL=(ALL,ALL) NOPASSWD: ALL" > /etc/sudoers.d/sles
]]>
          </source>
      </script>
    </chroot-scripts>
<init-scripts config:type="list">
      $SNIPPET('spacewalk/minion_script')
    </init-scripts>
</scripts>
</profile>
----

== Manage a {caasp} Cluster With {productname}

You can use {productname} to manage one or more existing {caasp} clusters.

[NOTE]
====
Only {caasp}{nbsp}4 is currently supported.
====

Before you begin, ensure you have installed your {caasp} cluster.

> You will need to manually copy some configuration information from your cluster to the {productname} Server:

Reviewer 3: The configuration should be copied to the elected Management Node.

Author: Fixed

> * Copy the ``skuba`` configuration directory from your cluster to the {productname} Server, and take a note of the file location.

Reviewer 3: I think this section should have a subtitle, something like `Prepare Management Node`.

Author: Fixed

. Click btn:[Save] and apply the highstate.

[NOTE]
====
You will not be able to use the management node until the highstate has been completed.
====

List all known clusters by navigating to menu:Clusters[Overview].
This list displays all existing clusters, along with the cluster type, and which management node they are associated with.
It also shows the nodes within the cluster, if the nodes are registered to {productname}.
For the nodes within a cluster, additional information from ``skuba`` and the {k8s} API are shown, including the role, status, and whether any updates are available.

For more information about the data available for nodes, see https://documentation.suse.com/suse-caasp/4/html/caasp-admin/_cluster_updates.html.

=== Manage Clusters

To manage a cluster in {productname}, add the cluster in the {webui}.

.Procedure: Adding an Existing Cluster
. In the {productname} {webui}, navigate to menu:Clusters[Overview] and click btn:[FIXME].
. Follow the prompts to provide information about your cluster, including the cluster type, and select the management node to associate.
. Type the path to the ``skuba`` configuration file for the cluster.
// For example?

> . Type the passwordless SSH key you want to use, or to the ``ssh-agent`` socket.

Reviewer 3: Maybe it would be useful to add a note here on how to forward the `ssh-agent`, something like:

There are two ways of using the ``ssh-agent`` with {caasp}:
1. Start the ``ssh-agent`` locally: ``eval $(ssh-agent)`` and add the ssh key: ``ssh-add <key>``. The socket used to access the agent is available in the ``$SSH_AGENT`` environment variable. 
2. Forward the ``ssh-agent`` to the management node from another machine: ``ssh -A <management node>``. The socket path is also available in the ``$SSH_AGENT`` environment variable. 

In both cases the path of the socket changes every time a new ``ssh-agent`` is started or a new ``ssh -A`` connection is started. 

The ``ssh-agent`` socket path can be updated at any time from the UI. The socket path can also be overridden when starting any cluster action that requires SSH access.

Author: Added

. Click btn:[FIXME].

For each cluster you manage with {productname}, a corresponding system group is created.
By default, the system group is called ``Cluster <cluster_name>``.
Refresh the system group to update the list of nodes.
Only nodes known to {productname} are shown.

You can remove clusters from {productname} by navigating to menu:Clusters[Overview], unchecking the cluster to be deleted, and clicking btn:[Delete Cluster].

[IMPORTANT]
====
Deleting a cluster removes the cluster from {productname}, it does not delete the cluster nodes.
Workloads running on the cluster will continue uninterrupted.
====

=== Manage Nodes

When you have the cluster created in {productname}, you can manage nodes within the cluster.

Before you add a new node to the cluster, check the management node can access the node you want to add using passwordless SSH, or the ``ssh-agent`` socket you are forwarding.

You also need to ensure that the node you want to add is registered to {productname}, and has a {caasp} channel assigned.

.Procedure: Adding a Node to a Cluster
. In the {productname} {webui}, navigate to menu:Clusters[Overview] and click btn:[Join Node].

> . Select the node to add from a list of available nodes.

Reviewer 3: multiple nodes can be selected, so I would say `node or nodes`

Author: Fixed

The list of available nodes includes only nodes that are registered to {productname}, are not management nodes, and are not currently part of any cluster.
. Follow the prompts to enter the {caasp} parameters for the node to be added.
. OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the nodes that are being added.
. Click btn:[Save] to schedule an action to add the node.
During this action, {productname} prepares the node for joining by disabling swap, then joins the node to the cluster.

.Procedure: Removing a Node from a Cluster

> . In the {productname} {webui}, navigate to menu:Clusters[Overview], uncheck the node to remove, and click btn:[Remove Node].

Reviewer 3: `check the node or nodes to remove`

Author: Fixed

{productname} will not stop you from issuing these operations if the system is not locked.
====

> == Autoinstallation profile of a {caasp}{nbsp}4 node

Reviewer 2: ```suggestion
== Autoinstallation Profile of a {caasp}{nbsp}4 Node
```

> {caasp}{nbsp}4 provides an AutoYaST profile to autoinstall a node in the `patterns-caasp-Management` package (see https://documentation.suse.com/suse-caasp/4.2/single-html/caasp-deployment/#_autoyast_preparation).
> The following is the {caasp}{nbsp}4 template customized to make use of {productname} channels and minion registration script that you can use as a starting point to automatically install a {caasp}{nbsp}4 using the Autoinstallation feature of {productname}.

Reviewer 2: ```suggestion
{caasp}{nbsp}4 provides an AutoYaST profile that you can use to autoinstall a node.
The profile is in the ``patterns-caasp-Management`` package.
For more information about the profile, see https://documentation.suse.com/suse-caasp/4.2/single-html/caasp-deployment/#_autoyast_preparation.

The template in this example is based on the {caasp}{nbsp}4 template, customized to make use of {productname}.
It includes channels and client registration scripts to install a {caasp}{nbsp}4 cluster with {productname} autoinstallation.
This script is intended for you to use as a basis for your own autoinstallation scripts.
```

Reviewer 2: Yes, I know I made it longer. You had a lot of information packed into those two sentences!

> .Example: Script for autoinstallation of a {caasp}{nbsp}4 node

Reviewer 2: This is quite lengthy, can we split it out to a sub-section?

Author: Fixed

For more information about the data available for nodes, see https://documentation.suse.com/suse-caasp/4/html/caasp-admin/_cluster_updates.html.

> .Procedure: Prepare the Configuration Credentials to the Management Node
> 
> You will need to prepare the configuration from your cluster to the management node:
> 
> . Copy the ``skuba`` configuration directory from your cluster to the {productname} Server, and take a note of the file location.
> // Default file location? --LKB 2020-06-04
> This is the directory that the ``skuba`` service creates after the cluster has been bootstrapped.
> 
> . Alternatively:
>   * Copy the passwordless private SSH key used to access the cluster nodes to the {productname} Server, and take a note of the file location.
> You need the current keys, and keys for any clients that you want to use in the future.
>   * You can use an ``ssh-agent`` socket, and provide the path to the socket when setting up the cluster. There are two ways of using the ``ssh-agent`` with {caasp}:
> 
>     ** By using ``ssh-agent`` locally:
>     *** Start the ssh-agent locally: ``eval $(ssh-agent)``
>     *** Add the SSH key: ``ssh-add <key>``
>     *** The socket used to access the agent is available in the ``$SSH_AGENT`` environment variable.
> 
>     ** Forward the `ssh-agent` to the management node from another machine:
> 
>       *** From your source machine: ``ssh -A <management node>``. The socket path is also available in the ``$SSH_AGENT`` environment variable.

Reviewer 2: This isn't really a procedure, it's three alternative options, so let's phrase it like that:

```suggestion
You need to prepare the configuration between the cluster and the management node:
There are three ways you can achieve this, choose the method that best suits your environment:

* Copy the ``skuba`` configuration directory from your cluster to the {productname} Server.
This is the directory that the ``skuba`` service creates after the cluster has been bootstrapped.
Take a note of the new file location for adding the cluster in the {productname} {webui}.
* Copy the passwordless private SSH key used to access the cluster nodes to the {productname} Server.
You need the current keys, and keys for any clients that you want to use in the future.
Take a note of the new file location for adding the cluster in the {productname} {webui}.
* Use an ``ssh-agent`` socket, and provide the path to the socket when setting up the cluster. 

If you choose to use an ``ssh-agent`` socket, there are two methods:
To use ``ssh-agent`` locally, start the ``ssh-agent`` on the local client with [command]``eval $(ssh-agent)``, and add the SSH key with [command]``ssh-add <key>``.
The socket used to access the agent is in the ``$SSH_AGENT`` environment variable.

Alternatively, forward the ``ssh-agent`` to the management node from another machine.
From your source machine use [command]``ssh -A <management node>``.
The socket used to access the agent is in the ``$SSH_AGENT`` environment variable.
```

Author: Fixed

> [NOTE]
> ====
> If you are using the ``ssh-agent`` method, the path of the socket changes every time a new ``ssh-agent``` is started or a new ``ssh -A`` connection is started. The ``ssh-agent`` socket path can be updated at any time from the UI of {productname}. The socket path can also be overridden when starting any cluster action that requires SSH access.
> ====

Reviewer 2: ```suggestion
[NOTE]
====
If you are using the ``ssh-agent`` method, the path of the socket changes every time a new ``ssh-agent``` is started or a new ``ssh -A`` connection is started. 
The ``ssh-agent`` socket path can be updated at any time from the {productname} {webui}. 
The socket path can also be overridden when starting any cluster action that requires SSH access.
====
```

You also need to ensure that the node you want to add is registered to {productname}, and has a {caasp} channel assigned.

> .Procedure: Adding Nodes(s) to a Cluster

Reviewer 2: ```suggestion
.Procedure: Adding Nodes to a Cluster
```

. In the {productname} {webui}, navigate to menu:Clusters[Overview] and click btn:[Join Node].

> . Select the node(s) to add from a list of available nodes.

Reviewer 2: ```suggestion
. Select the nodes to add from the list of available nodes.
```

The list of available nodes includes only nodes that are registered to {productname}, are not management nodes, and are not currently part of any cluster.
. Follow the prompts to enter the {caasp} parameters for the node(s) to be added.
. OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the node(s) that are being added.
. Click btn:[Save] to schedule an action to add the node(s).

> During this action, {productname} prepares the node(s) for joining by disabling swap, then joins the node(s) to the cluster.

Reviewer 2: Generally speaking, if you can use singular or plural, then we stick with plural. Adding the s in parentheses makes the doc harder to read and translate.

Author: Fixed

> . Follow the prompts to enter the {caasp} parameters for the node(s) to be added.
> . OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the node(s) that are being added.
> . Click btn:[Save] to schedule an action to add the node(s).
> During this action, {productname} prepares the node(s) for joining by disabling swap, then joins the node(s) to the cluster.

Reviewer 2: ```suggestion
. Follow the prompts to enter the {caasp} parameters for the nodes to be added.
. OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the nodes that are being added.
. Click btn:[Save] to schedule an action to add the nodes.
During this action, {productname} prepares the nodes for joining by disabling swap, then joins the nodes to the cluster.
```

Author: Fixed

> .Procedure: Removing Node(s) from a Cluster
> . In the {productname} {webui}, navigate to menu:Clusters[Overview], uncheck the node(s) to remove, and click btn:[Remove Node].
> . Follow the prompts to define the parameters for the node(s) to be removed.
> . OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the node(s) that are being removed.
> . Click btn:[Save] to schedule an action to remove the node(s).

Reviewer 2: ```suggestion
.Procedure: Removing Nodes from a Cluster
. In the {productname} {webui}, navigate to menu:Clusters[Overview], uncheck the nodes to remove, and click btn:[Remove Node].
. Follow the prompts to define the parameters for the nodes to be removed.
. OPTIONAL: Specify a custom ``ssh-agent`` socket that is valid only for the nodes that are being removed.
. Click btn:[Save] to schedule an action to remove the nodes.
```

Author: Fixed
